main.js

```diff
@@ -194,18 +203,81 @@ function createWindow () {
   });
 }
 
+function createTrayIcon() {
+  logger.info('Creating tray icon');
+
+  // TODO: better to not use a hardcoded icon file
+  tray = new Tray(path.join(__dirname, '/images/icon_48.png'));
+
+  trayContextMenu = Menu.buildFromTemplate([
+      {id: 'show', label: locale.messages.show.message, visible: !mainWindow.isVisible(), click: () => {showWindow();}},
+      {id: 'hide', label: locale.messages.hide.message, visible: mainWindow.isVisible(), click: () => {hideWindow();}},
+      {id: 'quit', label: locale.messages.quit.message, click: () => {app.quit();}}
+  ]);
+
+  // NOTE: a function equivalent to the following will become part of the
+  // prototype of Menu in a future release of electron (atm it is in 1.8.1-beta),
+  // so this definition can be removed in the future, when that version of electron
+  // will be released and stable. For details, please see:
+  // https://github.com/electron/electron/releases/tag/v1.8.1
+  // https://github.com/electron/electron/pull/10618
+  trayContextMenu.getMenuItemById = function (id) {
+    return this.items.find(item => item.id === id) || null;
+  }
+
+  tray.setContextMenu(trayContextMenu);
+
+  tray.on('clicked', () => {toggleWindowVisibility();});
+  tray.on('click', () => {toggleWindowVisibility();});
+
+  tray.setToolTip('Signal Desktop');
+}
+
 function showDebugLog() {
   if (mainWindow) {
     mainWindow.webContents.send('debug-log')
   }
 }
 
+function setTrayContextMenuItemsVisibility() {
+  if (tray && trayContextMenu) {
+    trayContextMenu.getMenuItemById('show').visible = !mainWindow.isVisible();
+    trayContextMenu.getMenuItemById('hide').visible = mainWindow.isVisible();
+    tray.setContextMenu(trayContextMenu); // need to re-set it on Linux
+  }
+}
+
 function showWindow() {
   if (mainWindow) {
     mainWindow.show();
+
+    // trick to bring above other windows and get focus
```

Comment:
Tell me more about this trick. What platforms has it been tested on (we've had particular difficulty with window placement/focus on windows) and where did it come from?

Also, note the 'show-window' IPC handler just a few lines above this. The right thing here might be to extract that out into a `focusWindow` method which shows the window if it isn't already showing. Then the tray handlers can use that instead of the simple `showWindow` method.

Reply:
I moved this in a specific method for the tray.
Sorry, I should definitely have given more detail here. I was aware that on some versions of GNOME there were problems when using this API, and the window may not be restored on top if it was not focused when it was hidden (cfr. also this issue on an app that uses the same API: https://github.com/Enrico204/Whatsapp-Desktop/issues/29 ), so I included it to be on the safe side (tested on Linux, KDE Plasma 5.11 and GNOME 3.26), even if I suspect it may be even unnecessary.